### PR TITLE
PUBDEV-8261: Added loss function equations for each distribution to Docs [nocheck]

### PR DESCRIPTION
**Pull request by hannah-tillman**
For: P[UBDEV-8261](https://h2oai.atlassian.net/browse/PUBDEV-8261)

I added the equation grid above the example, but let me know if you want it elsewhere!

### CODE REVIEW

File: h2o-docs/src/product/data-science/algo-params/distribution.rst [hunk 1]
```diff
@@ -80,6 +80,50 @@ Some examples of response distributions are provided below.
...
+|                   | - :math:`y<=f`:                                                                           |
+|                   | - :math:`2w*(\mid{y-f}\mid - \text{Huber_Delta})*\text{Huber_Delta}` (Equivalent to wMAE) |
++-------------------+-------------------------------------------------------------------------------------------+
+| Modified_Huber    | - with :math:`yf=(2y-1)*f`                                                                |
+|                   | - :math:`yf<-1`:                                                                          |
+|                   | - :math:`-4w*(yf)`                                                                        |
+|                   | - :math:`yf>1` : :math:`0`                                                                |
+|                   | - else: :math:`w*(yf)^2`                                                                  |
++-------------------+-------------------------------------------------------------------------------------------+
+| Tweedie           | with :math:`f=\text{link}(f)`, requires :math:`1>\text{_tweediePower}<2`:                 |
```
Comment (maurever) [inline h2o-docs/src/product/data-science/algo-params/distribution.rst:121]:
```suggestion
| Tweedie           | with :math:`f=\text{link}(f)`, requires :math:`1>\text{Tweedie_Power}<2`:                 |
```

Just to be consistent. :)

File: h2o-docs/src/product/data-science/algo-params/distribution.rst [hunk 2]
```diff
@@ -80,6 +80,50 @@ Some examples of response distributions are provided below.
...
+|                   | :math:`2w*y^{(2-\text{Tweedie_Power})/((1-\text{Tweedie_Power})*(2-\text{Tweedie_Power})) |
+|                   | -y*\text{exp}(f*(1-\text{Tweedie_Power}))/(1-\text{Tweedie_Power})                        |
+|                   | +\text{exp}(f*(2-\text{Tweedie_Power}))/(2-\text{Tweedie_Power})}`                        |
++-------------------+-------------------------------------------------------------------------------------------+
+
```
Comment (maurever) [inline h2o-docs/src/product/data-science/algo-params/distribution.rst:126]: We should also explain what `y`, `f,` and `w` mean.

y = true response
f = predicted response
w = weight

With distributions also link functions are connected. We should also mention this link functions and link inverse functions: 

- Identity: Gaussian, Huber, Laplace, Quantile

link(f) = f
linkInversion(f) = f

- Log: Multinomial, Poisson, Gamma, Tweede

link(f) = log(f)
linkInversion(f) = exp(f)

- Logit: Bernoulli, Quasibinomial, Modified Huber

link(f) = log(f / (1 - f)
linkInv(f) = 1 / (1 + exp(-f))

There are other link functions that are not associated with concrete distributions but can be used with a custom distribution: 

- Inverse
 
link(f) = 1/f (where f = f < 0 ? min(-1e-5, f) : max(-1e-5, f))
linkInversion(f) = link(f)

- Ologit

link(f) = log(f / (1 - f)
linkInversion(f) = 1 / (1 + exp(-f))

- Ologlog

link(f) = log(-1 * log(1-f) );
linkInversion(f) =1 - exp(-1 * exp(f))

- Oprobit

link(f) = normalDistribution.inverseCumulativeProbability(f)
linkInversion(f) = normalDistribution.cumulativeProbability(f)

@hannah-tillman, let me know if something is not clear to you. :)

**Review by maurever — CHANGES_REQUESTED** [review]

File: h2o-docs/src/product/data-science/algo-params/distribution.rst [hunk 3]
```diff
@@ -80,6 +80,87 @@ Some examples of response distributions are provided below.
...
+| Poisson           | - With :math:`f=link(f)`                                                                  |
+|                   | - :math:`-2w*(y*f-exp(f))`                                                                |
++-------------------+-------------------------------------------------------------------------------------------+
+| Gamma             | - With :math:`f=link(f)`                                                                  |
+|                   | - :math:`2w*(y*exp(-f)+f)`                                                                |
++-------------------+-------------------------------------------------------------------------------------------+
+| Laplace           | :math:`w*\mid{y-f}\mid`                                                                   |
++-------------------+-------------------------------------------------------------------------------------------+
+| Quantile          | - :math:`y>f: w*QuantileAlpha*(y-f)`                                                      |
+|                   | - :math:`y<=f: w*1-QuantileAlpha*(f-y)`                                                   |
```
Comment (neemah2o) [inline h2o-docs/src/product/data-science/algo-params/distribution.rst:114]: Should have parentheses around 1-QuantilAlpha
`w*(1-QuantileAlpha)*(f-y)`
From https://github.com/h2oai/h2o-3/blob/c395633dd20346bded1da3f87cc17f4d6b977a84/h2o-core/src/main/java/hex/DistributionFactory.java#L430

File: h2o-docs/src/product/data-science/algo-params/distribution.rst [hunk 4]
```diff
@@ -80,6 +80,87 @@ Some examples of response distributions are provided below.
...
++-------------------+-------------------------------------------------------------------------------------------+
+| Huber             | - :math:`\mid{y-f}\mid <= \text{Huber_Delta}`:                                            |
+|                   | - :math:`w*(y-f)^2` (Equivalent to wMSE)                                                  |
+|                   | - :math:`y<=f`:                                                                           |
+|                   | - :math:`2w*(\mid{y-f}\mid - \text{Huber_Delta})*\text{Huber_Delta}` (Equivalent to wMAE) |
++-------------------+-------------------------------------------------------------------------------------------+
+| Modified_Huber    | - with :math:`yf=(2y-1)*f`                                                                |
```
Comment (neemah2o) [inline h2o-docs/src/product/data-science/algo-params/distribution.rst:121]: I don't really like this term `yf`. It can confuse you with y*f. I would prefer giving it a new name. @maurever can this have a designated name? Does it have a mathematical name for this? Or can we just call it a new letter?

File: h2o-docs/src/product/data-science/algo-params/distribution.rst [hunk 5]
```diff
@@ -80,6 +80,87 @@ Some examples of response distributions are provided below.
...
+|                   | - :math:`yf<-1`:                                                                          |
+|                   | - :math:`-4w*(yf)`                                                                        |
+|                   | - :math:`yf>1` : :math:`0`                                                                |
+|                   | - else: :math:`w*(yf)^2`                                                                  |
++-------------------+-------------------------------------------------------------------------------------------+
+| Tweedie           | with :math:`f=\text{link}(f)`, requires :math:`1>\text{Tweedie_Power}<2`:                 |
+|                   | :math:`2w*y^{(2-\text{Tweedie_Power})/((1-\text{Tweedie_Power})*(2-\text{Tweedie_Power})) |
+|                   | -y*\text{exp}(f*(1-\text{Tweedie_Power}))/(1-\text{Tweedie_Power})                        |
+|                   | +\text{exp}(f*(2-\text{Tweedie_Power}))/(2-\text{Tweedie_Power})}`                        |
++-------------------+-------------------------------------------------------------------------------------------+
```
Comment (neemah2o) [inline h2o-docs/src/product/data-science/algo-params/distribution.rst:131]: Tweedie formating looks off when I view it here: https://github.com/h2oai/h2o-3/blob/5f6197eba3a69b0fc0a0e142272e7bcbb48ec331/h2o-docs/src/product/data-science/algo-params/distribution.rst
image

Not sure if this going to be fixed.

File: h2o-docs/src/product/data-science/algo-params/distribution.rst [hunk 6]
```diff
@@ -80,6 +80,87 @@ Some examples of response distributions are provided below.
...
++-------------------+---------------------------------------------------------------------------+
+| Log: Multinomial, | - :math:`link(f)=log(f)`                                                  |
+| Poisson, Gamma,   | - :math:`linkInversion(f)=exp(f)`                                         |
+| Tweedie           |                                                                           |
++-------------------+---------------------------------------------------------------------------+
+| Logit: Bernoulli, | - :math:`link(f)=(f/(f-1))`                                               |
+| Quasibinomial,    | - :math:`linkInversion(f)=1/(1+exp(-f))`                                  |
+| Modified_Huber    |                                                                           |
++-------------------+---------------------------------------------------------------------------+
+| Inversion         | - :math:`link(f) = 1/f (where f = f < 0 ? min(-1e-5, f) : max(-1e-5, f))` |
```
Comment (neemah2o) [inline h2o-docs/src/product/data-science/algo-params/distribution.rst:149]: I think most users won't be familiar with Java syntax. 

I would reformat it to:
(where if f < 0: f = min(-1e-5, f), else: f = max(-1e-5, f))
@maurever , am I interpreting this right?

**Review by neemah2o — CHANGES_REQUESTED** [review]

File: h2o-docs/src/product/data-science/algo-params/distribution.rst (hunk 5 above)
Comment (hannah-tillman) [reply h2o-docs/src/product/data-science/algo-params/distribution.rst:131]: Here is how the Tweedie formatting looks when the page is built. Does it still look off?

https://user-images.githubusercontent.com/52463461/129743156-456893d1-01bc-4367-94c7-570ea18b1a7f.mov

Comment (neemah2o) [reply h2o-docs/src/product/data-science/algo-params/distribution.rst:131]: For first part, It should be "requires 1> _tweediePower > 2 ", not "requires 1> _tweediePower < 2" 
The .mov shows everything in the exponent. But I think it should have three parts: 2w*y^(blah) -y*exp(blah)+exp(blah). For lines below, first line needs to end with } and remove last } on third line. Like this: 
`2w*y^{(2-\text{Tweedie_Power})/((1-\text{Tweedie_Power})*(2-\text{Tweedie_Power}))} |
|                   | -y*\text{exp}(f*(1-\text{Tweedie_Power}))/(1-\text{Tweedie_Power})                        |
|                   | +\text{exp}(f*(2-\text{Tweedie_Power}))/(2-\text{Tweedie_Power})`

File: h2o-docs/src/product/data-science/algo-params/distribution.rst [hunk 7]
```diff
@@ -80,6 +80,87 @@ Some examples of response distributions are provided below.
...
+| Quasibinomial     | - :math:`y==f:0`                                                                          |
+|                   | - :math:`f>1: -2w*y*log(f)`                                                               |
+|                   | - :math:`f<0: -2w*(1-y)*log(1-f) \text{ else}`                                            |
+|                   | - :math:`-2w*(y*log(f)+(1-y)*log(1-f))`                                                   |
++-------------------+-------------------------------------------------------------------------------------------+
+| Gaussian          | - Equivalent to wMSE (weighted mean squared error):                                       |
+|                   | - :math:`w*(y-f)^2`                                                                       |
++-------------------+-------------------------------------------------------------------------------------------+
+| Poisson           | - With :math:`f=link(f)`                                                                  |
+|                   | - :math:`-2w*(y*f-exp(f))`                                                                |
```
[the remark this answers is not in the document]
Comment (maurever) [reply h2o-docs/src/product/data-science/algo-params/distribution.rst:106]: Yes, it is more readable. The `f_link` sounds good.

File: h2o-docs/src/product/data-science/algo-params/distribution.rst (hunk 4 above)
Comment (maurever) [reply h2o-docs/src/product/data-science/algo-params/distribution.rst:121]: Good point. I am not sure it has some mathematical name. We can call it a new letter, for example, `z`...

File: h2o-docs/src/product/data-science/algo-params/distribution.rst (hunk 6 above)
Comment (maurever) [reply h2o-docs/src/product/data-science/algo-params/distribution.rst:149]: Yes, you are right, @neemah2o, the syntax is not readable. I just rewrite it from code to help @hannah-tillman to interpret it into the doc. There is also the possibility to use the `cases` formula:

```
.. math::
    f =
    \begin{cases}
        min(-1e-5, f) & \text{for }f<0 \\
        max(-1e-5, f)& \text{for }f>0 \\
    \end{cases}
```

**Review by maurever — CHANGES_REQUESTED** [review]
@hannah-tillman, please look at how to use math syntax in our doc. There are the same rules as for LaTeX I think: https://en.wikibooks.org/wiki/LaTeX/Mathematics

Usually, the `*` symbol is not used for multiplication. Also function as logarithm or exponential you can write as `\log(x)` or `\exp(x)`, etc. 

Somewhere you have` \text{log}(x) `somewhere only `log(x)`, correct is `\log(x)`.

My previous comment about link functions was not prepared to be directly copy-pasted, I just quickly extracted information from code for you. ;)

File: h2o-docs/src/product/data-science/algo-params/distribution.rst (hunk 5 above)
Comment (hannah-tillman) [reply h2o-docs/src/product/data-science/algo-params/distribution.rst:131]: With the update:

https://user-images.githubusercontent.com/52463461/129941718-38d2a9bb-6951-4c39-b68b-c3ddb213d69d.mov

**Review by maurever — APPROVED** [review]
Thank you, @hannah-tillman, for all changes you applied. I have one last suggestion. Otherwise, it looks good to me.

**Comment by hannah-tillman** [comment]
> Thank you, @hannah-tillman, for all changes you applied. I have one last suggestion. Otherwise, it looks good to me.

Thank you for all the help! :D

**Review by neemah2o — CHANGES_REQUESTED** [review]
A few more changes. Everything else looks good

File: h2o-docs/src/product/data-science/algo-params/distribution.rst [hunk 8]
```diff
@@ -80,6 +80,86 @@ Some examples of response distributions are provided below.
...
+| Gaussian          | - Equivalent to wMSE (weighted mean squared error):                                       |
+|                   | - :math:`w \times(y-f)^2`                                                                 |
++-------------------+-------------------------------------------------------------------------------------------+
+| Poisson           | - With :math:`f\text{_}link=link(f)`                                                      |
+|                   | - :math:`-2\times w\times(y\times f\text{_}link- \exp(f\text{_}link))`                    |
++-------------------+-------------------------------------------------------------------------------------------+
+| Gamma             | - With :math:`f\text{_}link=link(f)`                                                      |
+|                   | - :math:`2\times w\times (y\times \exp(-f\text{_}link)+f\text{_}link)`                    |
++-------------------+-------------------------------------------------------------------------------------------+
+| Laplace           | :math:`w\times\mid{y-f}\mid`                                                              |
```
[the remark this answers is not in the document]
Comment (hannah-tillman) [reply h2o-docs/src/product/data-science/algo-params/distribution.rst:111]: Just adding a space in the code unfortunately doesn't add a space in the equation. I added a space with the `text{}` function, however. It looks a little messy in the code but looks good in the printout:
Screen Shot 2021-08-31 at 9 12 20 PM

File: h2o-docs/src/product/data-science/algo-params/distribution.rst [hunk 9]
```diff
@@ -80,6 +80,86 @@ Some examples of response distributions are provided below.
...
++-------------------+---------------------------------------------------------------------------+
+| Log: Multinomial, | - :math:`link(f)=\log(f)`                                                 |
+| Poisson, Gamma,   | - :math:`linkInversion(f)= \exp(f)`                                       |
+| Tweedie           |                                                                           |
++-------------------+---------------------------------------------------------------------------+
+| Logit: Bernoulli, | - :math:`link(f)=(f/(f-1))`                                               |
+| Quasibinomial,    | - :math:`linkInversion(f)=1/(1+ \exp(-f))`                                |
+| Modified_Huber    |                                                                           |
++-------------------+---------------------------------------------------------------------------+
+| Inversion         | :math:`link\text{_}f= linkInversion\text{_}f = \begin{cases}min(-1e-5, f) |
```
Comment (maurever) [inline h2o-docs/src/product/data-science/algo-params/distribution.rst:148]: The style should be the same as in previous eq:
```suggestion
| Inversion         | :math:`link(f)= linkInversion(f) = \begin{cases}min(-1e-5, f) |
```

**Review by maurever — CHANGES_REQUESTED** [review]

**Review by neemah2o — APPROVED** [review]

**Review by maurever — APPROVED** [review]
Thank you @hannah-tillman!
